sslyze/scanner.py

```diff
@@ -80,30 +80,43 @@ def __post_init__(self) -> None:
                 raise ValueError(f"Received an extra argument for a scan command that wasn't enabled: {scan_command}")
 
 
-# TypedDict for simpler/matching JSON output and makes fetching a field easier
-class ScanCommandResultsDict(TypedDict, total=False):
+@dataclass(frozen=True)
+class ScanCommandResults:
     """A dictionary of results for every scan command that was scheduled against a specific server.
     """
 
     # Field is present if the corresponding scan command was scheduled and was run successfully
-    certificate_info: CertificateInfoScanResult
-    ssl_2_0_cipher_suites: CipherSuitesScanResult
-    ssl_3_0_cipher_suites: CipherSuitesScanResult
-    tls_1_0_cipher_suites: CipherSuitesScanResult
-    tls_1_1_cipher_suites: CipherSuitesScanResult
-    tls_1_2_cipher_suites: CipherSuitesScanResult
-    tls_1_3_cipher_suites: CipherSuitesScanResult
-    tls_compression: CompressionScanResult
-    tls_1_3_early_data: EarlyDataScanResult
-    openssl_ccs_injection: OpenSslCcsInjectionScanResult
-    tls_fallback_scsv: FallbackScsvScanResult
-    heartbleed: HeartbleedScanResult
-    robot: RobotScanResult
-    session_renegotiation: SessionRenegotiationScanResult
-    session_resumption: SessionResumptionSupportScanResult
-    session_resumption_rate: SessionResumptionRateScanResult
-    http_headers: HttpHeadersScanResult
-    elliptic_curves: SupportedEllipticCurvesScanResult
+    certificate_info: Optional[CertificateInfoScanResult] = None
+    ssl_2_0_cipher_suites: Optional[CipherSuitesScanResult] = None
+    ssl_3_0_cipher_suites: Optional[CipherSuitesScanResult] = None
+    tls_1_0_cipher_suites: Optional[CipherSuitesScanResult] = None
+    tls_1_1_cipher_suites: Optional[CipherSuitesScanResult] = None
+    tls_1_2_cipher_suites: Optional[CipherSuitesScanResult] = None
+    tls_1_3_cipher_suites: Optional[CipherSuitesScanResult] = None
+    tls_compression: Optional[CompressionScanResult] = None
+    tls_1_3_early_data: Optional[EarlyDataScanResult] = None
+    openssl_ccs_injection: Optional[OpenSslCcsInjectionScanResult] = None
+    tls_fallback_scsv: Optional[FallbackScsvScanResult] = None
+    heartbleed: Optional[HeartbleedScanResult] = None
+    robot: Optional[RobotScanResult] = None
+    session_renegotiation: Optional[SessionRenegotiationScanResult] = None
+    session_resumption: Optional[SessionResumptionSupportScanResult] = None
+    session_resumption_rate: Optional[SessionResumptionRateScanResult] = None
+    http_headers: Optional[HttpHeadersScanResult] = None
+    elliptic_curves: Optional[SupportedEllipticCurvesScanResult] = None
+
+    def keys(self) -> Iterator[str]:
+        for field_ in fields(self):
+            key = field_.name
+            if getattr(self, key) is not None:
+                yield key
+
+    def items(self) -> Iterator[Tuple[str, Any]]:
+        for key in self.keys():
+            yield key, getattr(self, key)
+
+    def __len__(self) -> int:
+        return len(list(self.keys()))
```

Comment:
The ``keys``, ``items`` and ``__len__`` functions are mainly for ease of use with the current implementation.

Reply:
Also you don't need to rely on getattr() to fetch a result; you can just access the field directly.

Reply:
I also thought about it again. And clearly there is two "variants of access" for this dataclass:
1. direct access to one field (can be done by ``server_scan_results.field_name`` or ``getattr(server_scan_results, field_name_var)``
1. iterating over/counting/using all fields (sometimes only the keys).

As we only want to count/iterate over the fields which are not ``None`` in the second case, some convenience wrapper/functions are needed imho (as you do not want to implement this logic multiple times). One can this/discuss about the names (which are very dict-like right now; I happily change them, if you have any preferences), but I think we need them, as this dataclass should be somehow "iterable".

> Also you don't need to rely on getattr() to fetch a result; you can just access the field directly.

I'm not aware, how to access a dataclass field whose name I have in a variable without ``getattr()``?! Please review the code again and give me hint how to do this if its possible.